README.md
# fatca-ides-php

[![Build Status](https://travis-ci.org/shadiakiki1986/fatca-ides-php.svg?branch=master)](http://travis-ci.org/shadiakiki1986/fatca-ides-php)

[packagist](https://packagist.org/packages/shadiakiki1986/fatca-ides-php)

This is a PHP library that converts bank client data to FATCA files submittable via the IDES gateway.
If you find in this library any code that is specific to the Financial Institution for which I did this work,
do not hesitate to point them out to me by opening an [issue](https://github.com/shadiakiki1986/fatca-ides-php/issues).

For more information check the [IRS FATCA IDES Technical FAQs](http://www.irs.gov/Businesses/Corporations/FATCA-IDES-Technical-FAQs)

For other language implementations, please check the [IRS github page](https://github.com/IRSgov)
 

# Pre-requisites
* php
* client data in php array form
* SSL certificate for your financial institution
* Private and public keys used to get the SSL certificate

# Installation instructions
Install [composer](https://getcomposer.org/download/)
```bash
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php -r "if (hash_file('SHA384', 'composer-setup.php') === '55d6ead61b29c7bdee5cccfb50076874187bd9f21f65d8991d46ec5cc90518f447387fb9f76ebae1fbbacf329e583e30') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;"
php composer-setup.php
php -r "unlink('composer-setup.php');"
```

Install php extensions
```bash
sudo apt-get install php7.0-xml php-mbstring php-zip php-mcrypt php-pear libyaml-dev php-dev php-bcmath
sudo pecl install yaml-2.0.0
sudo echo "extension=yaml.so" > /etc/php/7.0/mods-available/yaml.ini
sudo ln -s /etc/php/7.0/mods-available/yaml.ini /etc/php/7.0/cli/conf.d/20-yaml.ini

composer require shadiakiki1986/fatca-ides-php
composer require swiftmailer/swiftmailer # To enable sending emails
composer install
[sudo] apt-get install php5-mcrypt
[sudo] php5enmod mcrypt
[sudo] service apache2 restart # needed for web applications served with apache
```

If emails are enabled, pass your config (similar to [this](http://symfony.com/doc/current/cookbook/email/email.html#configuration) ) to the `toEmail` function of `Transmitter` class (check example below)

Next, download the financial institution's ssl certificate, private key, and public key.
Note that the public key can be extracted from the certificate (on the TODO to be implied without the need for the user to specify it)

# Examples
For a complete example of how to use this library, please check [IDES-Data-Preparation-Php](https://github.com/shadiakiki1986/IDES-Data-Preparation-Php).

## Example 1: generate ZIP file for submission on the IDES gateway

```php
require __DIR__.'/vendor/autoload.php';
use FatcaIdesPhp\Transmitter;

// set tax year
$taxYear=2014;

// example client data
$di=array(
  array("Compte"=>"1234","ENT_FIRSTNAME"=>"Clyde","ENT_LASTNAME"=>"Barrow","ENT_FATCA_ID"=>"123-1234-123","ENT_ADDRESS"=>"Some street somewhere","ResidenceCountry"=>"US","posCur"=>100000000,"cur"=>"USD","ENT_TYPE"=>"Individual"),
  array("Compte"=>"5678","ENT_FIRSTNAME"=>"Bonnie","ENT_LASTNAME"=>"Parker","ENT_FATCA_ID"=>"456-1234-123","ENT_ADDRESS"=>"Dallas, Texas","ResidenceCountry"=>"US","posCur"=>100,"cur"=>"LBP","ENT_TYPE"=>"Individual")
);

// config with paths to files downloaded
$config = array(
  # SSL certificate bought using the private key
  "FatcaCrt" => "/var/lib/IDES/keys/ssl_certificate.crt",
  # Private key used to get the SSL certificate
  "FatcaKeyPrivate" => "/var/lib/IDES/keys/institution-fatca-private.pem",
  # public key extracted out of private key above
  "FatcaKeyPublic" => "/var/lib/IDES/keys/institution-fatca-public.pem",

  # Fatca Sender GIIN
  "ffaid" => 'A1BBCD.00000.XY.123',
  # Fatca Receiver GIIN, e.g. IRS
  "ffaidReceiver" => '000000.00000.TA.840',

  # email configuration if Transmitter::toEmail is used
  "swiftmailer" => array(
    "host" => "my.host.com",
    "port" => 12345,
    "username" => "myusername",
    "password" => "mypassword",
    "name" => "My Name",
    "reply" => "myemail@host.com")

);

$factory = new \FatcaIdesPhp\Factory();

$fda = new \FatcaIdesPhp\FatcaDataArray($di);
$fdo = $factory->array2oecd($fda); // , false, null, $taxYear, "zip", "", $config
$tmtr=$factory->transmitter($fdo);
$tmtr->getZip();
```

## Example 2: To submit a correction XML file
Use the same code as above, but use
```php
$sxe = simplexml_load_file("path/to/xml");
$fdx = new \FatcaIdesPhp\FatcaDataXml($sxe);
$tmtr=$factory->transmitter($fdx);
```
Also, check the [correction.php](https://github.com/shadiakiki1986/IDES-Data-Preparation-Php/blob/master/www/correction.php) command-line utility from the `IDES-Data-Preparation-PHP` project


# License
Please check [[LICENSE]]

# Testing
```bash
composer install
composer run-script test
```

# Schema versions
## Current state
The schema files and IRS public key are committed to this repo in `assets`.

The committed versions are the 2.0 versions.

This package was tested succesfully against the IDES test gateway on Jan 31, 2017.
The test included
* submitting test data to the server via SFTP using the [IDES-Data-Preparation-Php](https://github.com/shadiakiki1986/IDES-Data-Preparation-Php) package
* receiving an ACK for the submission via email
* receiving a response from the IRS using the [IDES-Data-Preparation-Php](https://github.com/shadiakiki1986/IDES-Data-Preparation-Php) package

## Updating

To update them, run the following two scripts:
1. `./assets/update.sh`: Downloads files from IRS website
  * Note that the URL's in the script may need to be updated because newer versions are usually uploaded to new endpoints
  * Note also that there are some file renaming lines in the script so that I can diff versions
    * e.g. `git diff ba722d8bcda61f657529a67cdbec873a29dc7d70 5f9545b565ddf0d41997b29c704c3990813f4bb8` will diff version 1.1 and 2.0

2. `php updateXsd.php`: converts the schema files to PHP classes in `src/FatcaXsdPhp`
  * I moderate the updates and commit them to the repo if suitable
  * Open issues on these are:
    * I currently fix the below issues manually after the update
    * There seems to be a problem in `src/FatcaXsdPhp/oecd/ties/stffatcatypes/v1/Address_Type.php` by having two AddressFree fields
    * `@xmlNamespace urn:oecd:ties:fatca:v1` is missing from `src/FatcaXsdPhp/FATCA_OECD` pending https://github.com/moyarada/XSD-to-PHP/issues/36
    * The AddressFree field has the wrong namespace: should be urn:oecd:ties:stffatcatypes:v1 instead of `urn:oecd:ties:fatca:v1`
    * ReportingGroup.php is not generated + its `@var` is missing
    * `src/FatcaXsdPhp/oecd/ties/stffatcatypes/v1/NamePerson_Type.php`
      * @xmlNamespace urn:oecd:ties:fatca:v1
      * changed manually to
      * @xmlNamespace urn:oecd:ties:stffatcatypes:v1

## Manual validation
The validation of an existing xml file can be done manually against the schema on ubuntu using
```bash
sudo apt-get install libxml2-utils
xmllint --noout --schema assets/fatcaxml/FatcaXML.xsd path/to/file.xml
```


CHANGELOG.md
Version 0.3.3 (2019-03-20)
* updated the IRS Public Key from their website

Version 0.3.2 (2019-02-26)
* remove condition that disallows "individual" account with "substantial owner" (similar to POA)

2016-06-08
* moving code out of (IDES-Data-Preparation-Php)[https://github.com/shadiakiki1986/IDES-Data-Preparation-Php] to here
 * check there for earlier repo commits

2016-06-07
* converting the whole folder to a proper php package installable via composer
* splitting the sample web app into its own folder

2016-03-17
* converted scripts in tests to phpunit tests
 * these should be factored out to test the code in the Receiver and Transmitter classes instead of a copy of the code in the tests
* added some variable checking for the config
* minor documentation improvements

2015-07-02
* after a long struggle to understand why our submission was yielding a security threat alert, I learned that the way I was signing the XML file was wrong.
* Correcting it resulted in a successful submission
* I cleaned up the code a bit before making this commit. I probably should have made the messy commit right after the successful submission, and then made a 2nd cleaned up commit... too late now
* The people from lbi.fatca.ides@irs.gov were the ones who pointed out that my problem was with the signing

2015-06-26
* refactored `ENT_COD` to Compte and accountsTotalUsd to posCur
* had forgotten to include currency .. fixed

2015-06-12
* attempting at integrating correction doc ref ID, but documentation isn''t so clear
* added taxYear to getFatcaData and interface

2015-06-06
* data source functionality moved to separate function lib/getFatcaData.php
* refactored var/www/api/getFatcaClients.php to var/www/api/transmitter.php
* cleaned up code a bit

2015-06-04
* dropped ''xsi:schemaLocation'' which seems to have been causing ''virus threat''
* generating new ID''s instead of static one in MessageRefId
* setting timezone to UTC
* DocRefID: currently just putting ''Ref ID123''
* getting balances from Bankflow instead of dummy 0''s
* skipping some accounts just for testing purposes

2015-06-03
* testing in 2nd window
* got feedback saying that decrypted contents contain a virus threat
* working on the issue
* replaced xml signing with package xmlseclibs instead of custom function
* added class to parse received zip into xml
* replacing invalid characters

2015-05
* released first draft that successfully uploads


src/FatcaIdesPhp/Array2Oecd.php
<?php

namespace FatcaIdesPhp;

// Converter from FatcaDataArray to FatcaDataOecd
class Array2Oecd {

  function __construct($fda) {
    assert($fda instanceof FatcaDataArray);
    $this->fda=$fda;
  }

  function getMessageSpec() {
    $ms = new \oecd\ties\stffatcatypes\v2\MessageSpec_Type();
    $ms->SendingCompanyIN = $this->fda->conMan->config["ffaid"];
    $ms->TransmittingCountry = "LB";
    $ms->ReceivingCountry = "US";
    $ms->MessageType = "FATCA";
    $ms->MessageRefId = $this->fda->guidManager->get();

    $ms->ReportingPeriod = sprintf("%s-12-31",$this->fda->taxYear);
    $ms->Timestamp = $this->fda->ts3;
    return $ms;
  }

  function getReportingFI() {
    $rfi = new \FatcaXsdPhp\CorrectableOrganisationParty_Type();
    $rfi->Name = new \oecd\ties\stffatcatypes\v1\NameOrganisation_Type();
    $rfi->Name->value = "FFA Private Bank";
    $rfi->Address = new \oecd\ties\stffatcatypes\v2\Address_Type();
    $rfi->Address->CountryCode = new \oecd\ties\isofatcatypes\v1\CountryCode_Type();
    $rfi->Address->CountryCode->value = "LB";
    $rfi->Address->AddressFree = "Foch street";

    $rfi->TIN = $this->getTin(
      $this->fda->conMan->config["ffaid"],
      "US");

    $rfi->DocSpec = new \FatcaXsdPhp\DocSpec_Type();
    $rfi->DocSpec->DocTypeIndic=$this->fda->docType;
    $rfi->DocSpec->DocRefId=$this->fda->guidManager->get();
    // corrDocRefId ...
    return $rfi;
  }

  function getAccountHolder($x) {
      $ah = new \FatcaXsdPhp\AccountHolder_Type();
      switch($x["ENT_TYPE"]) {
        case "Individual":
          $ah->Individual = $this->getIndividual($x);
          break;
        case "Corporate":
          $ah->AcctHolderType = new \FatcaXsdPhp\FatcaAcctHolderType_EnumType();
          $ah->AcctHolderType->value = "FATCA102";
          $ah->Organisation = $this->getOrganisation($x);
          break;
        default:
          throw new \Exception("Unsupported ENT_TYPE '".$x["ENT_TYPE"]."'");
      }
      return $ah;
  }

  function getTin($tin,$issuer) {
    $TIN = new \oecd\ties\stffatcatypes\v2\TIN_Type();
    $TIN->value=$tin;
    $TIN->issuedBy=is_null($issuer)?"US":$issuer;
    return $TIN;
  }

  function getTinWrapper($x) {
    return $this->getTin(
      Utils::cleanTin($x['ENT_FATCA_ID']),
      array_key_exists("ENT_FATCA_ID_ISSUER",$x)?$x["ENT_FATCA_ID_ISSUER"]:null);
  }

  function getIndividual($x) {
    $ind = new \oecd\ties\stffatcatypes\v2\PersonParty_Type();
    $ind->TIN = $this->getTinWrapper($x);

    $ind->Name = new \oecd\ties\stffatcatypes\v2\NamePerson_Type();
    $ind->Name->FirstName = $x['ENT_FIRSTNAME'];
    $ind->Name->LastName = $x['ENT_LASTNAME'];
    $ind->Address = new \oecd\ties\stffatcatypes\v2\Address_Type();
    $ind->Address->CountryCode = $x['ResidenceCountry'];
    $ind->Address->AddressFree = Utils::cleanAddress($x['ENT_ADDRESS']);
    return $ind;
  }

  function getOrganisation($x) {
    $org = new \oecd\ties\stffatcatypes\v2\OrganisationParty_Type();
    $org->Name = new \oecd\ties\stffatcatypes\v1\NameOrganisation_Type();
    $org->Name->value = $x['ENT_FIRSTNAME'];
    $org->Address = new \oecd\ties\stffatcatypes\v2\Address_Type();
    $org->Address->CountryCode = $x['ResidenceCountry'];
    $org->Address->AddressFree = Utils::cleanAddress($x['ENT_ADDRESS']);
    $org->TIN = $this->getTinWrapper($x);

    return $org;
  }

  function getPayment_Type($type,$currCode,$value) {
    $pay = new \FatcaXsdPhp\Payment_Type();
    $pay->Type = $type;
    $pay->PaymentAmnt = new \oecd\ties\stffatcatypes\v1\MonAmnt_Type();
    $pay->PaymentAmnt->currCode = $currCode;
    $pay->PaymentAmnt->value = $value;
    return $pay;
  }

  function getPayments($x) {
      $payments=array();
      if(array_key_exists('dvdCur',$x)) {
        $pay = $this->getPayment_Type(
          'FATCA501',
          $x['cur'],
          floor($x['dvdCur']));
        array_push($payments,$pay);
      }

      if(array_key_exists('intCur',$x)) {
        $pay = $this->getPayment_Type(
          'FATCA502',
          $x['cur'],
          floor($x['intCur']));
        array_push($payments,$pay);
      }
      return $payments;
  }

  function convert() {
    // convert to FATCA_OECD and return FatcaDataOecd
    // much of this is inspired from fatca-ides-php/src/FatcaIdesPhp/FataDataArray
    $this->oecd=new \FatcaXsdPhp\FATCA_OECD();
    $this->oecd->version="1.1";
    $this->oecd->MessageSpec = $this->getMessageSpec();

    $this->oecd->FATCA=new \FatcaXsdPhp\Fatca_Type();
    $this->oecd->FATCA->ReportingFI = $this->getReportingFI();

    $this->oecd->FATCA->ReportingGroup = new \FatcaXsdPhp\ReportingGroup();

    // gather account reports
    $accountReports = array();
    foreach($this->fda->data as $x) {
      $ar = new \FatcaXsdPhp\CorrectableAccountReport_Type();
      $ar->DocSpec = new \FatcaXsdPhp\DocSpec_Type();
      $ar->DocSpec->DocTypeIndic=$this->fda->docType;
      $ar->DocSpec->DocRefId=$this->fda->guidManager->get();

      $ar->AccountNumber = $x['Compte'];

      $ar->AccountHolder = $this->getAccountHolder($x);

      if(array_key_exists("SubstantialOwner",$x)) {
        if($x["ENT_TYPE"]!="Corporate") throw new \Exception("Cannot have type Individual and substantial owners for: ".$x["Compte"]);

        $substOwns = array();
        foreach($x["SubstantialOwner"] as $so) {
          $subst = new \FatcaXsdPhp\SubstantialOwner_Type();
          $subst->Individual = $this->getIndividual($so);
          array_push($substOwns,$subst);
        }
        if(count($substOwns)>0) $ar->SubstantialOwner = $substOwns;
      }


      $ar->AccountBalance = new \oecd\ties\stffatcatypes\v1\MonAmnt_Type();
      $ar->AccountBalance->currCode = $x['cur'];
      $ar->AccountBalance->value = $x['posCur'];

      $payments=$this->getPayments($x);
      if(count($payments)>0) $ar->Payment = $payments;
      array_push($accountReports,$ar);
   }

    if(count($accountReports)>0) $this->oecd->FATCA->ReportingGroup->AccountReport = $accountReports;

    $this->fdo = new FatcaDataOecd($this->oecd);
  }

} //end class
